server/index.js

```diff
@@ -53,6 +59,92 @@ app.get('/api/products/:productId', (req, res, next) => {
     .catch(err => next(err));
 });
 
+app.post('/api/cartItems', (req, res, next) => {
+  const { productId, customizations, quantity } = req.body;
+  const qtyNum = parseInt(quantity, 10);
+  const prodNum = parseInt(productId, 10);
+  if (!productId) {
+    throw new ClientError(400, 'productId are required.');
+  } else if (!Number.isInteger(qtyNum) || !Number.isInteger(prodNum)) {
+    throw new ClientError(400, 'quantity and productId need to be positive integers.');
+  }
+  const cartToken = req.headers['x-access-token'];
+  if (cartToken === 'null') {
+    const sql = `
+    insert into "carts"
+         values (default)
+      returning *
+      `;
+    db.query(sql)
+      .then(result => {
+        const payload = { cartId: result.rows[0].cartId };
+        const token = jwt.sign(payload, process.env.TOKEN_SECRET);
+        const response = { payload, token };
+        return response;
+      })
+      .then(response => {
+        const newCartId = response.payload.cartId;
+        const newToken = response.token;
+        const sql = `
+    insert into "cartItems" ("productId", "customizations", "quantity", "cartId")
+         values ($1, $2, $3, $4)
+      returning *
+      `;
+        const params = [productId, customizations, quantity, newCartId];
+        db.query(sql, params)
+          .then(result => {
+            const cartDetails = result.rows[0];
+            res.status(200).json({ cartDetails, newToken });
+          })
+          .catch(err => next(err));
+      });
+  } else {
+    const payload = jwt.verify(cartToken, process.env.TOKEN_SECRET);
+    const cartId = payload.cartId;
+    const sqlCart = `
+      select "productId",
+             "customizations",
+             "cartItemsId",
+             "quantity"
+        from "cartItems"
+       where "productId" = $1;
+       `;
+    const params = [productId];
+    db.query(sqlCart, params)
+      .then(result => {
+        const newItem = Object.values(customizations);
+        const numQuantity = JSON.parse(quantity);
+        for (let i = 0; i < result.rows.length; i++) {
+          const cartItemId = result.rows[i].cartItemsId;
+          const cartQuantity = result.rows[i].quantity;
+          const cartItem = Object.values(result.rows[i].customizations);
+          if (equal(cartItem, newItem)) {
+            const newQuantity = numQuantity + cartQuantity;
+            const sqlItem = `
+              update "cartItems"
+                 set "quantity" = $1
+               where "cartItemsId" = $2
+               returning *;
+              `;
+            const params = [newQuantity, cartItemId];
+            return db.query(sqlItem, params);
```

Comment:
You can deduplicate this code by dividing up your work into two steps
1. Get/decode the token or create a cart and token
2. Add the cartItem

This can also be simplified by _always_ responding with the token. You'll need some conditional logic to accomplish this. But it helps that if your `return` from a `.then()` callback, the returned value is passed into the next `.then()` callback. Here is some pseudocode

```js
Promise
  .resolve()
  .then(() => {
    // if there is a token, decode it and return an object containing token and cartId
    // otherwise, return db.query(createCart).then(() => create a token and return an object containing token and cartId)
  .then(cartIdAndToken => {
    // now you have a cartId and Token, the previous .then() made sure that they exist
    // return addCartItem().then(() => res.json(theTokenAndCartItem))
  })
  .catch(err => next(err)) // only one .catch needed

```